src/main/java/org/jenkinsci/plugins/scriptsecurity/sandbox/whitelists/ProxyWhitelist.java

```diff
@@ -51,16 +87,37 @@
     /** anything wrapping us, so that we can propagate {@link #reset} calls up the chain */
     private final Map<ProxyWhitelist,Void> wrappers = new WeakHashMap<ProxyWhitelist,Void>();
 
+    /** Caches valid permission checks or false if no {@link CacheableWhitelist} approves the entry.
+     *  This allows us to avoid redundant checks against the {@link Whitelist}s that statically approve
+     *   on the basis of the Method/Field/Constructor signature ONLY (see contract for CacheableWhitelist).
+     *
+     *  Synchronization not required because we already synchronize on delegates.
+     */
+    private HashMap<String, Boolean> permittedCache = new HashMap<String, Boolean>();
+
     public ProxyWhitelist(Collection<? extends Whitelist> delegates) {
         reset(delegates);
     }
 
+    /** Sorter that ensures we check {@link CacheableWhitelist}s before the dynamically-evaluating whitelists such as
+     *  {@link AclAwareWhitelist}. */
+    static final Comparator<Whitelist> CACHEABLE_WHITELIST_FIRST = new Comparator<Whitelist>() {
+        @Override
+        public int compare(Whitelist o1, Whitelist o2) {
+            if (o1 instanceof CacheableWhitelist ^ o2 instanceof CacheableWhitelist) {  // If they don't match
+                return (o1 instanceof CacheableWhitelist) ? -1 : 1;
+            }
+            return 0;
```

Comment:
Maybe more readable as

```java
return o1 instanceof CacheableWhitelist == o2 instanceof CacheableWhitelist ? 0 : o1 instanceof CacheableWhitelist ? -1 : 1;
```

Reply:
What, no love for clever (ab)use of operators?